src/ir/auth_logic/auth_logic_ast_traversing_visitor.h

```diff
@@ -0,0 +1,322 @@
+  }
+  // Invoked before all the children of `query` are visited.
+  virtual Result PreVisit(CopyConst<IsConst, Query>& query) {
+    return GetDefaultValue();
+  }
+  // Invoked after all the children of `query` are visited.
+  virtual Result PostVisit(CopyConst<IsConst, Query>& query,
+                           Result in_order_result) {
+    return in_order_result;
+  }
+  // Invoked before all the children of `program` are visited.
+  virtual Result PreVisit(CopyConst<IsConst, Program>& program) {
+    return GetDefaultValue();
+  }
+  // Invoked after all the children of `program` are visited.
+  virtual Result PostVisit(CopyConst<IsConst, Program>& program,
+                           Result in_order_result) {
+    return in_order_result;
+  }
+
+  // TODO (#644) aferr
+  // The Visits for the Datalog IR classes (RelationDeclaration, Predciate)
+  // are here temporarily until these AST classes are refactored out
+  // of the Datalog IR.
+
+  virtual Result Visit(
+      CopyConst<IsConst, datalog::RelationDeclaration>& relationDeclaration) {
+    return GetDefaultValue();
+  }
+
+  virtual Result Visit(CopyConst<IsConst, datalog::Predicate>& predicate) {
+    return GetDefaultValue();
+  }
+
+  // The remaining Visits are meant to follow the convention
+
+  Result Visit(CopyConst<IsConst, Principal>& principal) final override {
+    Result pre_visit_result = PreVisit(principal);
+    return PostVisit(principal, std::move(pre_visit_result));
+  }
+
+  Result Visit(CopyConst<IsConst, Attribute>& attribute) final override {
+    Result pre_visit_result = PreVisit(attribute);
+    Result fold_result =
+        FoldResult(FoldResult(std::move(pre_visit_result),
+                              attribute.principal().Accept(*this)),
+                   // TODO(#644 aferr): fix this to use predicate().Accept once
+                   // predicate has been refactored into ast.h
+                   Visit(attribute.predicate()));
+    return PostVisit(attribute, std::move(fold_result));
+  }
+
+  Result Visit(CopyConst<IsConst, CanActAs>& canActAs) final override {
+    Result pre_visit_result = PreVisit(canActAs);
+    Result fold_result =
+        FoldResult(FoldResult(std::move(pre_visit_result),
+                              canActAs.left_principal().Accept(*this)),
+                   canActAs.right_principal().Accept(*this));
+    return PostVisit(canActAs, std::move(fold_result));
+  }
+
+  Result Visit(CopyConst<IsConst, BaseFact>& baseFact) final override {
+    Result pre_visit_result = PreVisit(baseFact);
+    Result variant_visit_result = std::visit(
+        raksha::utils::overloaded{
+            [this](const datalog::Predicate& pred) {
+              return VariantVisit(pred);
+            },
+            [this](const Attribute& attrib) { return VariantVisit(attrib); },
+            [this](const CanActAs& canActAs) {
+              return VariantVisit(canActAs);
+            }},
+        baseFact.GetValue());
```

Comment:
That doesn't quite work unfortunately... I tried what you suggested tweaking it slightly to the following because it's missing a param and a capture to the following:

```
    Result variant_visit_result = std::visit([*this](const auto &branch) {
      return VariantVisit(branch);
    }, base_fact.GetValue());
```

and auto is not able to deduce the type:
```
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/variant:548:12: error: cannot deduce return type 'auto' from returned value of type '<overloaded function type>'
```

along with this other cryptic error I did not dig into:
```
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/variant:659:14: error: attempt to use a deleted function
      return _VSTD::__invoke_constexpr(_VSTD::forward<_Visitor>(__visitor),
```

 I remember running into a similar issue when implementing `lowering_ast_datalog`.

I think actually the better way to remove this code would be to remove the use of `variant`s entirely from the implementation of the AST and replace them with sublcasses representing the different forms. This visitor is already implementing the pattern in a way that uses dynamic dispatch and inheritance to call the `Accept` implementation specific to the subclasses of the nodes. If you think that makes sense, I'll open an issue. Otherwise, it's left as-is.

Reply:
Sounds good, if we're already using subclasses and vtables to subtype, might as well use them everywhere. I think that's a good way forward. Unfortunate that the generic lambda didn't work :(